### PR TITLE
Fix sending duration with HTTPS proxies

### LINKED ISSUE #6317: Duration metric is not accurate for "double TLS" MITM proxies (edge case)
### Brief summary

I'm writing this up to document an edge case where K6 http request duration metrics are not accurate. This isn't something that will affect many people, but it's here mostly for Google or anyone stumbling upon it.

I have been using K6 to test a MITM proxy that is called via TLS, and then talks upstream via TLS. Consider an application that receives an environment variable `HTTPS_PROXY=https://foo...`. There are two TLS handshakes here: between the client and proxy, and between the client and upstream.

Image

The problem is this breaks assumptions made by K6's metrics:

- `blocked` covers the period from `getConn` to `gotConn`. It includes both TLS handshakes.
- `duration` covers the period from the _first_ TLSHandshakeDone to the end of the request

The issue is, because there are two TLS Handshakes, this means the `blocked` period and `duration` overlap. This inflates the assumed overall time of the request, because instead of duration strictly being `sending + waiting + receiving`, it includes the second TLS step as well

In other words, the second TLS negotiation is effectively counted twice in the metrics

This is a fairly unusual setup and ultimately this isn't going to be a problem for my tests. However I think it's worth flagging up in case anyone is seeing a TLS double-count or mismatching values with a MITM SSL proxy.

The actual solution might be as simple as to only count `duration` as starting from `GotConn`

### k6 version

1.8.1

### OS

Ubuntu 26.04

### Docker version and image (if applicable)

_No response_

### Steps to reproduce the problem

- Run K6 with HTTPS_PROXY env var set to a https-transport based proxy
- Make connections with no connection reuse (TLS renegotiation each time)
- Observe how duration + blocked is larger than total time of request to complete

### Expected behaviour

- Blocked + duration should cover the whole request time

### Actual behaviour

- Second TLS negotiation time is double-counted

### PR DESCRIPTION
**Pull request by kausthubhk**
## What?

Fixes request timing metrics when an HTTPS proxy is used with an HTTPS target. Sending is now calculated from GotConn to WroteRequest, preventing it from overlapping with Blocked.

Unavailable or out-of-order timestamps produce a zero Sending duration. A deterministic regression test covers the double-TLS event ordering.

## Why?

When both the proxy connection and upstream connection use TLS, the previous Sending start time could be recorded before GotConn. This caused Sending to include time already accounted for by Blocked, producing incorrect request timing metrics.

## Testing

- go test ./lib/netext/httpext/

## Related issue

Closes #6317

### CODE REVIEW

File: lib/netext/httpext/tracer.go [hunk 1]
```diff
@@ -343,21 +343,10 @@ func (t *Tracer) Done() *Trail {
 	if tlsHandshakeDone != 0 && tlsHandshakeStart != 0 {
 		trail.TLSHandshaking = time.Duration(tlsHandshakeDone - tlsHandshakeStart)
 	}
+	if gotConn != 0 && wroteRequest > gotConn {
+		trail.Sending = time.Duration(wroteRequest - gotConn)
+	}
```
Comment (szkiba) [inline lib/netext/httpext/tracer.go:350]: Nit: the removed `switch` had per-case comments explaining why each timestamp was chosen (TLS vs plain vs the HTTP/2 quirk). Now that it's collapsed into one branch, could you add a short comment here explaining why `gotConn` is the right universal reference point? e.g.:

```go
// Sending starts once the connection is actually obtained (gotConn),
// regardless of whether it came via a plain connect, a single TLS
// handshake, or a proxy tunnel with a nested upstream TLS handshake.
if gotConn != 0 && wroteRequest > gotConn {
    trail.Sending = time.Duration(wroteRequest - gotConn)
}
```
Would help future readers avoid re-litigating the TLS/proxy edge case this PR just fixed.

File: lib/netext/httpext/tracer_test.go [hunk 2]
```diff
@@ -236,10 +236,32 @@ func TestTracerNegativeHttpSendingValues(t *testing.T) {
 		builtinMetrics := metrics.RegisterBuiltinMetrics(registry)
 		trail.SaveSamples(builtinMetrics, &metrics.TagsAndMeta{Tags: registry.RootTagSet()})
 
-		require.True(t, trail.Sending > 0)
+		require.GreaterOrEqual(t, trail.Sending, time.Duration(0))
```
Comment (szkiba) [inline lib/netext/httpext/tracer_test.go:239]: Nit: this assertion went from `require.True(t, trail.Sending > 0)` to `require.GreaterOrEqual(t, trail.Sending, time.Duration(0))`. Could you confirm the reason for the relaxation? I traced through this test's write-fails-then-retry-on-new-connection scenario and `Sending` should still come out strictly positive with the new `gotConn`-based formula, so I want to make sure this is just guarding against timestamp-resolution flakiness rather than papering over a case where `Sending` now silently collapses to zero.

**Review by szkiba — DISMISSED** [review]

File: lib/netext/httpext/tracer_test.go (hunk 2 above)
Comment (kausthubhk) [reply lib/netext/httpext/tracer_test.go:239]: @szkiba 
Yes, this is only guarding against timestamp-resolution flakiness. On Windows I observed GotConn and WroteRequest being recorded in the same clock tick, so the correct GotConn-based duration is legitimately zero. The old TLS-based formula started from an earlier timestamp and therefore remained positive in that case. This test specifically protects the failed-write retry path from producing a negative duration. TestTracerSendingStartsAtGotConn separately uses distinct timestamps and verifies that the normal GotConn → WroteRequest interval is strictly positive.

File: lib/netext/httpext/tracer.go (hunk 1 above)
Comment (kausthubhk) [reply lib/netext/httpext/tracer.go:350]: Good point, added a comment. Thanks!

**Comment by kausthubhk** [comment]
Hi @szkiba  @inancgumus ,

I've addressed the review comments and also synced the branch with master. It looks like the current CI workflows are waiting for maintainer approval before they can run. When you get a chance, could you please approve the workflow runs and take another look at the PR? Happy to address anything else if needed. Thanks!

**Review by szkiba — APPROVED** [review]

**Review by inancgumus — APPROVED** [review]
